packages/components/src/utils/hooks/use-controlled-value.ts

```diff
@@ -0,0 +1,38 @@
+/**
+ * External dependencies
+ */
+import { isNil } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useState } from '@wordpress/element';
+
+type Props< T > = {
+	defaultValue?: T;
+	value?: T;
+	onChange?: ( value: T ) => void;
+};
+
+/**
+ * Simplified and improved implementation of useControlledState.
+ *
+ * @param  props
+ * @param  props.defaultValue
+ * @param  props.value
+ * @param  props.onChange
+ * @return The controlled value and the value setter.
+ */
+export function useControlledValue< T >( {
+	defaultValue,
+	onChange,
+	value: valueProp,
+}: Props< T > ): [ T | undefined, ( value: T ) => void ] {
+	const hasValue = ! isNil( valueProp );
```

Comment:
why check "nil" and not "undefined". For me "null" is a valid value meaning there's actually a value equal to "null"

Reply:
I tend to agree with Riad here but just wanted to bring up that if we use the idiomatic `value != null` we essentially achieve the same that `isNil()` does, but avoid unnecessary Lodash usage.

Reply:
`null` isn't a valid value for any controllable element as far as I know. Why should we allow it here?

You'll receive this warning if you try to pass `null` to an `input` for example:

```
"Warning: `value` prop on `input` should not be null. Consider using an empty string to clear the component or `undefined` for uncontrolled components.", 
```

Reply:
> null isn't a valid value for any controllable element as far as I know. Why should we allow it here?

I think it depends of what you consider a controllable element. If we restrict the definition to DOM element, yes, it's not a valid value but AFAIK, this hook is supposed to be used to make any component controlled? and for instance `value=null` seems value in a `Select` component that allows to select an object among a list of objects or none.

Reply:
Yes any controllable element, but I hadn't considered anything other inputs as controllable 🤔

Reply:
`null` actually isn't a valid `value` prop for an `option` element, if you pass `null` the value of the `select` wrapping it will be the text inside the `option`.

Reply:
I was not talking about the html `select` or `option`, more about a custom abstraction on top of that, that allows you to pick say a user object in a list of users. And if you'd want to make that component support both controlled and uncontrolled behavior, you could decide to wrap it with the hook from this PR. (Unless I misunderstood the purpose of that hook :) )

Reply:
Gotcha, I can't imagine a custom `Select` component that doesn't just run on `select` and `option` under the hood though... that seems like it would be a lot of over-engineering, right?

I'm happy to change the conditional if you feel strongly about it but testing it will be interesting 🙂

Reply:
> Gotcha, I can't imagine a custom Select component that doesn't just run on select and option under the hood though... that seems like it would be a lot of over-engineering, right?

I'm not saying it doesn't run the `select` under the hood. Let me try to clarify here. Imagine we have a controlled `Select` component that has the following API:


```is
const someArray = [ { id: 'id', label: 'something' } ];
const [ value, setValue ] = useState( null ); 
<Select value={ value } options={ someArray } onChange={ onChange } />
```

it allows you to select an object in a list or no object if the value is `null`, only supports a controlled mode though. (Notice that I didn't mention how the Select is implemented internally but it's most likely using select and options)

Now you want to make this component work in both controlled/uncontrolled mode, If I understand the proposed hook here properly, I should be able to do this:

```
const MyControlledUncontrolledSelect = ( props ) => {
   const [ value, onChange ] = useControlledValue( props );
   
   return <Select value={value} onChange={onChange} />;
}
```

and now I can use it however I want.

--- 
I don't feel strongly here personally but I feel that making the check `!==undefined` is a more generic behavior.

Reply:
Aha, thanks for the explanation! I'll make the change 👍